src/imap-login/imap-login-client.c

```diff
@@ -389,6 +391,52 @@ static void imap_client_create(struct client *client, void **other_sets)
 				   IMAP_LOGIN_MAX_LINE_LENGTH);
 	if (imap_client->set->imap_literal_minus)
 		imap_parser_enable_literal_minus(imap_client->parser);
+
+	/* create our capability list from CAPABILITY_BANNER_STRING */
+	imap_client->capability_list =
+		 imap_capability_list_create(NULL);
+
+	if (*imap_client->set->imap_capability == '\0')
+		imap_capability_list_append_string(imap_client->capability_list,
+				CAPABILITY_BANNER_STRING);
+	else if (*imap_client->set->imap_capability != '+') {
+		imap_capability_list_append_string(imap_client->capability_list,
+				imap_client->set->imap_capability);
+		explicit_capability = TRUE;
+	} else {
+		/* add the capability banner string to the cap list */
+		imap_capability_list_append_string(imap_client->capability_list,
+				CAPABILITY_BANNER_STRING);
+		/* add everything after the plus to our cap list */
+		imap_capability_list_append_string(imap_client->capability_list,
+				imap_client->set->imap_capability + 1);
+	}
+
+	/* add STARTTLS to cap list if it is needed */
+	if (strcmp(client->ssl_set->ssl, "no") != 0) {
+		imap_capability_list_add(imap_client->capability_list,
+				"STARTTLS", IMAP_CAP_VISIBILITY_REQUIRE_ALL |
```

Comment:
I think this now returns STARTTLS when connecting to imaps port. Also should likely use `client_is_tls_enabled(client)` instead of `strcmp(client->ssl_set->ssl, "no") != 0`

Reply:
> I think this now returns STARTTLS when connecting to imaps port.
Is there any changes I need to do regarding this?

Other than that, I have updated the strcmp to client_is_tls_enabled().

Reply:
Actually it seems to be a bit wrong. Anyway my main concern was that the original code was:

```
	if (client_is_tls_enabled(client) && !client->tls)
		str_append(cap_str, " STARTTLS");
```

But your code isn't using `client->tls` anywhere. It seems you're actually using `client->starttls`. So I think now if you connect to localhost the CAPABILITY list won't show STARTTLS?

Reply:
Hmmm... I think I understand what you're saying...

The STARTTLS capability is being registered with:

                   imap_capability_list_add(imap_client->capability_list, "STARTTLS",
                                            IMAP_CAP_VISIBILITY_FLAG_REQUIRE_ALL |
                                            IMAP_CAP_VISIBILITY_PREAUTH |
                                            IMAP_CAP_VISIBILITY_INSECURE);

This means that STARTTLS will be shown if the connection is INSECURE, and we are at PREAUTH.

We still want to display STARTTLS over secured connections like localhost, just not when TLS/SSL is already established in some form.

So would it work if we changed the registration to:

                   imap_capability_list_add(imap_client->capability_list, "STARTTLS",
                                            IMAP_CAP_VISIBILITY_FLAG_REQUIRE_ALL |
                                            IMAP_CAP_VISIBILITY_PREAUTH |
                                            IMAP_CAP_VISIBILITY_TLS_INACTIVE);

That way in the get_capability function we will have:

        /* is tls/ssl enabled? */
        if (client->tls)
                flags |= IMAP_CAP_VISIBILITY_TLS_ACTIVE;
        else
                flags |= IMAP_CAP_VISIBILITY_TLS_INACTIVE;

The above will effectively perform the `!client->tls` check from the original code, and then the other half of the check is performed at the time of registering the STARTTLS keyword:

        if (client_is_tls_enabled(client)) { 
                imap_capability_list_add(imap_client->capability_list, "STARTTLS",
                                         IMAP_CAP_VISIBILITY_FLAG_REQUIRE_ALL |
                                         IMAP_CAP_VISIBILITY_PREAUTH |
                                         IMAP_CAP_VISIBILITY_TLS_INACTIVE);
        }

So as long as TLS is enabled we will add STARTTLS to the capability list under the condition that it is only ever shown at: preauth when TLS/SSL is not active whatsoever.

Is this an accurate solution to the problem you are describing?

Reply:
Just wanted to follow up - haven't seen any response on the proposed solution - can you please let me know if what has been proposed is an accurate solution?

Thanks!